### PR TITLE
allow modification of Replica1 type to Replica2 or Replica3

### PR DESCRIPTION
**Pull request by amarts**
Signed-off-by: Amar Tumballi <amar@kadalu.io>

### CODE REVIEW

**Comment by amarts** [comment]
This should be enough to support 'distribute' to 'replicate' type of the volume. But keeping it as draft as not sure what are the checks/validations we need to handle to prevent other disastrous situation where users can spoil an healthy volume.

Also, need a mechanism to regenerate the volume files for the volume which changed, so running volumes in nodeplugin and other places can properly work.

**Comment by leelavg** [comment]
As discussed earlier on slack about changing Replica1 to ~other~ Replica2 before a migration, just curious are you aiming to facilitate replace-brick related to #537 with this PR?

**Comment by aravindavk** [comment]
> As discussed earlier on slack about changing Replica1 to ~other~ Replica2 before a migration, just curious are you aiming to facilitate replace-brick related to #537 with this PR?

This is add brick + change of Volume type.

**Comment by leelavg** [comment]
> This is add brick + change of Volume type.

Ack, what I was aiming at is, if we are able to change volume type if there's a possibility of changing back from Replica2 to Replica1 I can see a possibility of using it for migrating bricks.

Not to hijack this PR, I'll raise an RFE and then discuss on it's feasibility.

**Comment by amarts** [comment]
If someone has kadalu setup handy, can this branch be tested? 

How to test ?

Create a storage pool with just one brick, now modify the same storage pool to be of Replica3 type. The major check to do is to check if the mounts in csi pods are now consuming new volfile.

File: csi/main.py [hunk 1]
```diff
@@ -41,6 +43,21 @@ def mount_storage():
         logging.info(logf("Volume is mounted successfully", hvol=hvol))
     return
 
+def reconfigure_mounts():
```
Comment (leelavg) [inline csi/main.py]: Single handler expects a func should accept two args, can you pls change the signature to `reconfigure_mounts(*args)`?

File: csi/volumeutils.py [hunk 2]
```diff
@@ -824,6 +826,16 @@ def generate_client_volfile(volname):
     with open(info_file_path) as info_file:
         data = json.load(info_file)
 
+    if not VOL_DATA.get(volname, None):
```
Comment (leelavg) [inline csi/volumeutils.py:833]: If possible can you please add some comments for this snippet?

File: csi/watch-vol-changes.sh [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+CONFIG_MAP_INFO_DIR="/var/lib/gluster"
+
+CSI_PROCESS_ID=$(ps aux|grep python | awk '{ print $2}' | xargs);
```
Comment (leelavg) [inline csi/watch-vol-changes.sh]: Better grep for only `[m]ain.py`, `start.py` doesn't know how to handle `SIGHUP` and it'll restart whole pod.
Disregard if the intention is to restart all CSI pods.

Comment (amarts) [reply csi/watch-vol-changes.sh]: You may be right. Intention is to only send signal to `main.py`. Still testing. My local setup is still broken, and wanted to see what makes sense. Expect changes here.

**Comment by hunter86bg** [comment]
Can I use something like

 $ kubectl kadalu storage-add storage-pool-1 --type Replica3 \
     --path kube2:/bricks/brick2/brick --path kube3:/bricks/brick3/brick

to test replica1 to replica3 ?

File: csi/main.py (hunk 1 above)
Comment (amarts) [reply csi/main.py]: Ack

**Comment by leelavg** [comment]
> $ kubectl kadalu storage-add storage-pool-1 --type Replica3
> --path kube2:/bricks/brick2/brick --path kube3:/bricks/brick3/brick

@hunter86bg thanks for your interest. For testing the change first you need to pull changes and build containers  (`make build-containers`) locally if not doing any hacks.

Atm, Amar is making use of CI to come to a final conclusion so I'd suggest pls wait until code path is fixed.

**Comment by amarts** [comment]
Ok, my observation is that inotify is not working on configmap changes. So, need some other strategy.

**Comment by amarts** [comment]
As SIGHUP was not working properly, I took the little 'not-so-efficient' but works properly strategy of checking for hash value of the configmap file whenever a 'mount_glusterfs()' is called. This is not efficient, but adding few milliseconds to every PV create, and every management operation is OK IMO to be correct.

**Comment by leelavg** [comment]
After a bit of search about any existing issues with configMap and inotify, found [this](https://github.com/kubernetes/minikube/issues/1551#issuecomment-494957948)

**Comment by amarts** [comment]
Thanks for the link @leelavg ! It made sense... I didn't suspect minikube earlier. Good to know.

Regardless, the latest implementation is ready to try out separately! I am intending to keep both SIGHUP and reload volfile during every mount call in code for another few months so we can see which is working better.

File: csi/main.py [hunk 4]
```diff
@@ -42,6 +44,23 @@ def mount_storage():
...
+    for volume in host_volumes:
+        if volume["type"] == "External":
+            # Need to skip remount external
+            continue
+        hvol = volume["name"]
+        try:
+            if reload_glusterfs(volume):
+                logging.info(logf("Volume reloaded successfully", hvol=hvol))
+        except CommandException:
+            logging.error(logf("Unable to mount volume", hvol=hvol))
```
Comment (aravindavk) [inline csi/main.py:61]: Wrong error message

Comment (amarts) [reply csi/main.py:61]: Resolved in #555

**Comment by amarts** [comment]
Taken up in #555

**Pull request closed without merging**